rfc/state-model.rst

```diff
@@ -0,0 +1,299 @@
+
+During the course of an update, a specific firmware image can change from being *active* to *second*, or from *second* to *active*. For example:
+
+* An image will switch from being *second* --- while being prepared --- to *active* following installation.
+* An image will switch from being *active* to *second* when it becomes the backup image during installation of new firmware.
+
+Firmware Store state model
+--------------------------
+
+A persistent state model for an in-progress Firmware Update is required for two reasons:
+
+1. When updating firmware that is essential for system operation, a system restart is required to complete installation, and start executing the new system firmware. Communication between the runtime system and the Bootloader regarding the required installation operations must be done via memory.
+2. On constrained devices, the download of a new firmware image may take a long time, due to bandwidth or power limitations. A device restart should not invalidate the update progress that has already occurred.
+
+The basic flow required to update firmware within the constraints and meeting the requirements identified for the API is captured in the existing v0.7 document.
+
+For version 1.0, we propose to present the same operational flow as a state model of the **Firmware Store**, instaead of an **individual image**, and the existing APIs will work to cause transitions within this alternative state model.
+
+Note
+    Although readers might be familiar with the v0.7 image lifecycle-based state model, the definition of a Firmware Store-based state model is clearer without providing explicit linkage to the image states defined in v0.7.
+
+Firmware Store states
+~~~~~~~~~~~~~~~~~~~~~
+
+A proposed set of Firmware Store states is as follows:
+
+.. list-table::
+    :header-rows: 1
+
+    * - State
+      - Description
+
+    * - READY
+      - This is the normal state for the Store, where there is just one image, the *active*, which is currently in use by the system.
+
+        The Store is ready for a new firmware update to be started.
+
+    * - WRITING
+      - The Client is writing a new firmware image to the *second*, in preparation for installation.
+
+        When writing is complete, it can be installed.
+
+    * - STAGED
+      - Installation of the *second* has been requested, but the system must be restarted as the final update operation runs within the Bootloader.
+
+        This state is transient.
+
+    * - FAILED
+      - An installation of the *second* has been attempted, but has failed for some reason. The failure reason is recorded in the Store.
+
+        The *second* needs to be erased before another update can be attempted.
```

Comment:
Does it actually need to be erased? It may be a wast of power/time to actually erase the data. Marking the image as invalid may be enough?

Reply:
You don't strictly need to erase the second store to run the active one, but until you have done that you won't be able to run further firmware updates so can't go back to READY state.